logs: fix missing allocation logs after update to Nomad 1.5.4

tgross opened the pull request:
When the server restarts for the upgrade, it loads the `structs.Job` from the Raft snapshot/logs. The jobspec has long since been parsed, so none of the guards around the default value are in play. The empty field value for `Enabled` is the zero value, which is false.

This doesn't impact any running allocation because we don't replace running allocations when either the client or server restart. But as soon as any allocation gets rescheduled (ex. you drain all your clients during upgrades), it'll be using the `structs.Job` that the server has, which has `Enabled = false`, and logs will not be collected.

This changeset fixes the bug by adding a new field `Disabled` which defaults to false (so that the zero value works), and deprecates the old field.

Fixes #17076

schmichael commented on api/tasks.go:
```diff
@@ -639,16 +639,21 @@ func (g *TaskGroup) AddSpread(s *Spread) *TaskGroup {
+	MaxFiles      *int `mapstructure:"max_files" hcl:"max_files,optional"`
+	MaxFileSizeMB *int `mapstructure:"max_file_size" hcl:"max_file_size,optional"`
+
+	// COMPAT(1.7.0): Enabled had to be swapped for Disabled to fix a backwards
+	// compatibility bug when restoring pre-1.5.4 jobs. Remove in 1.7.0
+	Enabled *bool `mapstructure:"enabled" hcl:"enabled,optional"`
```
Can we remove in 1.6.0? I see this as less of a traditional deprecation and more of a bug removal that unfortunately needs to span >1 release.

I think it's safe to remove in 1.6.0 because:

1. Basically everyone who upgraded to 1.5.4 *must upgrade to 1.5.5* and therefore won't need this field in 1.6.0.
2. Anyone who didn't upgrade to 1.5.4 won't need this compat field in 1.6.0 anyway

(if somehow there are people who upgrade to 1.5.4 and don't care either way... i don't think they'll care about when this is removed either)

tgross replied:
Agreed. Will fix.

schmichael commented on command/agent/job_endpoint.go:
```diff
@@ -1809,11 +1805,23 @@ func apiLogConfigToStructs(in *api.LogConfig) *structs.LogConfig {
 	}
+	if in.Disabled == nil {
+		// COMPAT(1.7.0): fix for backwards compatibility
+		out.Disabled = !dereferenceBool(in.Enabled)
```
If `Enabled = nil` won't this propagate the bug we're trying to fix?

tgross replied:
Oops, I missed that because I didn't have a test for that case. The API struct gets canonicalized first so we don't hit it in practice, but we should belt-and-suspenders on this!

schmichael commented on jobspec/parse_task.go:
```diff
@@ -263,6 +263,7 @@ func parseTask(item *ast.ObjectItem, keys []string) (*api.Task, error) {
 			"max_files",
 			"max_file_size",
 			"enabled",
```
```suggestion
			"enabled", COMPAT(1.7.0): remove in favor of disabled
```

or 1.6 if we go with that

tgross commented:
@schmichael I've addressed your comments and added some extra tests around those cases. I've also done bench testing across upgrades: both from 1.5.3 to 1.5.5-dev and from 1.5.4 to 1.5.5-dev. The nice thing I discovered with the fix is that allocations with missing `logmon` will automatically get a logmon created when upgrading from 1.5.4 to 1.5.5-dev without having to reschedule the allocation. The allocrunner prestart hook does the right thing for us.
